pkg/remotecluster/remotecluster.go

```diff
@@ -0,0 +1,202 @@
+package remotecluster
+
+import (
+	"sync"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	v1net "k8s.io/api/networking/v1"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+)
+
+type RemoteCluster struct {
+	stopCh chan struct{}
+
+	ClusterID             string
+	ClientSet             kubernetes.Interface
+	podInformer           cache.SharedIndexInformer
+	networkPolicyInformer cache.SharedIndexInformer
+
+	eventChanMutex *sync.Mutex
+	eventChan      chan *Event
+}
+
+type EventType string
+
+const (
+	AddEvent    EventType = "Added"
+	DeleteEvent EventType = "Deleted"
+	UpdateEvent EventType = "Updated"
+)
+
+type ObjectType string
+
+const (
+	NetworkPolicy ObjectType = "np"
+	Pod           ObjectType = "pod"
+)
+
+type Event struct {
+	Cluster *RemoteCluster
+	Type    EventType
+	ObjType ObjectType
+	Objs    []interface{}
+}
+
+func New(clusterID string, clientSet kubernetes.Interface) *RemoteCluster {
+
+	factory := informers.NewSharedInformerFactory(clientSet, time.Hour*24)
+	podInformer := factory.Core().V1().Pods().Informer()
+	networkPolicyInformer := factory.Networking().V1().NetworkPolicies().Informer()
+
+	resourceWatcher := &RemoteCluster{
+		stopCh:                make(chan struct{}),
+		ClusterID:             clusterID,
+		ClientSet:             clientSet,
+		podInformer:           podInformer,
+		networkPolicyInformer: networkPolicyInformer,
+		eventChanMutex:        &sync.Mutex{},
+	}
+
+	podInformer.AddEventHandler(resourceWatcher)
+	networkPolicyInformer.AddEventHandler(resourceWatcher)
+
+	return resourceWatcher
+}
+
+func (rw *RemoteCluster) HasSynced() bool {
+	return rw.podInformer.HasSynced() &&
+		rw.networkPolicyInformer.HasSynced()
+}
+
+// Stop will stop the running informers
+func (rw *RemoteCluster) Stop() {
+	close(rw.stopCh)
+}
+
+func (rw *RemoteCluster) Stopped() bool {
+	select {
+	case _, ok := <-rw.stopCh:
+		// channel is not ok to read, means it's closed now
+		return !ok
+	default:
+		// channel was ok to read, but had no messages
+		return false
+	}
+}
+
+func (rw *RemoteCluster) Run(onSyncDoneFunc func(resourceWatcher *RemoteCluster)) {
+	go rw.podInformer.Run(rw.stopCh)
+	go rw.networkPolicyInformer.Run(rw.stopCh)
+
+	go func() {
```

Comment:
I disagree it's premature optimization, more like defensive optimization when you've seen things fail because of similar issues.

I'm not sure I will use it yet in the current implementation after all your comments though.